share/eds/store.go

```diff
@@ -197,31 +202,59 @@ func (s *Store) Blockstore() bstore.Blockstore {
 	return s.bs
 }
 
-// CARBlockstore returns the IPFS blockstore that provides access to the IPLD blocks stored in an
-// individual CAR file.
-func (s *Store) CARBlockstore(ctx context.Context, dataHash []byte) (dagstore.ReadBlockstore, error) {
-	key := shard.KeyFromString(fmt.Sprintf("%X", dataHash))
+// CARBlockstore returns an IPFS Blockstore providing access to individual shares/nodes of a
+// specific EDS identified by DataHash and registered on the Store. NOTE: The Blockstore does not
+// store whole Celestia Blocks but IPFS blocks. We represent `shares` and NMT Merkle proofs as IPFS
+// blocks and IPLD nodes so Bitswap can access those.
+func (s *Store) CARBlockstore(
+	ctx context.Context,
+	root share.DataHash,
+) (dagstore.ReadBlockstore, error) {
+	key := shard.KeyFromString(root.String())
+	accessor, err := s.getCachedAccessor(ctx, key)
+	if err != nil {
+		return nil, fmt.Errorf("eds/store: failed to get accessor: %w", err)
+	}
+	return accessor.bs, nil
+}
+
+// GetDAH returns the DataAvailabilityHeader for the EDS identified by DataHash.
+func (s *Store) GetDAH(ctx context.Context, root share.DataHash) (*share.Root, error) {
+	key := shard.KeyFromString(root.String())
 	accessor, err := s.getAccessor(ctx, key)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("eds/store: failed to get accessor: %w", err)
 	}
+	defer accessor.Close()
 
-	return accessor.bs, nil
+	carHeader, err := carv1.ReadHeader(bufio.NewReader(accessor))
+	if err != nil {
+		return nil, fmt.Errorf("eds/store: failed to read car header: %w", err)
+	}
+
+	dah := dahFromCARHeader(carHeader)
+	if !bytes.Equal(dah.Hash(), root) {
+		return nil, fmt.Errorf("eds/store: content integrity mismatch from CAR for root %x", root)
+	}
+	return dah, nil
 }
 
-func (s *Store) getAccessor(ctx context.Context, key shard.Key) (*accessorWithBlockstore, error) {
-	// try to fetch from cache
-	accessor, err := s.cache.Get(key)
-	if err != nil && err != errCacheMiss {
-		log.Errorw("unexpected error while reading key from bs cache %s: %s", key, err)
+// dahFromCARHeader returns the DataAvailabilityHeader stored in the CIDs of a CARv1 header.
+func dahFromCARHeader(carHeader *carv1.CarHeader) *header.DataAvailabilityHeader {
+	rootCount := len(carHeader.Roots)
+	rootBytes := make([][]byte, 0)
+	for _, root := range carHeader.Roots {
+		rootBytes = append(rootBytes, ipld.NamespacedSha256FromCID(root))
 	}
```

Comment:
you know how large rootBytes can be so allocate and copy into it

Reply:
The append style is nice, but you can set the capacity in advance after zero to preallocate

Reply:
oops yeah, forgot about 3 argument make. done